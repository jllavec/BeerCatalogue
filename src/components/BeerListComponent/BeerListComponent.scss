.beer-list {
  $parent: &;
  background-color: $bg-color-dark;
  border-radius: 8px;
  padding: rem(16);

  &--gallery {
    @include breakpoint(tablet) {
      display: grid;
      grid-gap: rem(16);
      grid-template-columns: repeat(auto-fill, 250px);
      justify-content: center;
      justify-items: center;
      width: 100%;
    }

    #{$parent}__element-fav {
      bottom: rem(8);
      top: unset;
    }
  }

  &__display-selector {
    display: flex;
    margin-bottom: rem(32);
  }

  &__gallery-trigger {
    display: none;

    @include breakpoint(tablet) {
      display: block;
    }
  }

  &__element {
    background-color: $bg-color;
    border-radius: 8px;
    cursor: pointer;
    display: flex;
    flex-direction: column;
    margin-bottom: rem(8);
    padding: rem(8);
    position: relative;
    width: 100%;

    &-title {
      font-size: rem(20);
    }

    &-emoji {
      margin-top: auto;
    }

    &-fav {
      bottom: rem(8);
      fill: #fff;
      position: absolute;
      right: rem(8);
      stroke: #000;
      width: rem(20);

      @include breakpoint(tablet) {
        bottom: unset;
        top: rem(8);
      }

      &--selected {
        fill: #0ececf;
      }
    }
  }
}