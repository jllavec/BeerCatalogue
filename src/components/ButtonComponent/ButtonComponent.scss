.button {
  background: $secondary-color;
  border: 2px solid $secondary-color;
  border-radius: 50px;
  box-shadow: 0px 4px 10px rgb(0 0 0 / 10%);
  color: $white;
  display: block;
  font-size: 14px;
  font-size: rem(18);
  font-style: normal;
  font-weight: bold;
  letter-spacing: 1px;
  letter-spacing: rem(0.15);
  line-height: rem(16);
  line-height: rem(22);
  min-width: rem(200);
  padding: rem(17) rem(25);
  text-align: center;
  transition: .3s all ease;

  &:hover {
    background: $bg-color-green;
    border: 2px solid $bg-color-green;
    color: $white;
  }

  &:disabled {
    background-color: grey;
  }
}
