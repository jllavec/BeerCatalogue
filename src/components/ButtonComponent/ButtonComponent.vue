<template>
  <button
    class="button"
    :disabled="disabled"
  >
    {{ label }}
  </button>
</template>

<script>
export default {
  name: 'ButtonComponent',

  props: {
    label: {
      type: String,
      default: ''
    },
    disabled: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="scss" scoped src="./ButtonComponent.scss"  />
