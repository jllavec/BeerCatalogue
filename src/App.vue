<template>
  <div class="app">
    <Header />
    <div class="container">
      <div
        v-if="!showFav"
        class="search"
      >
        <label for="Name">Search your beer</label>
        <input
          id="Name"
          v-model="search"
          type="text"
          name="Name"
          placeholder="Search by name / brew date (mm-yyyy)"
        >
        <p v-if="errorSearch">
          Invalid search params. Type any valid name or date in format (mm-yyyy)
        </p>
      </div>
      <ListComponent
        :beer-list="beerListShow"
        @displayModal="triggerModal"
        @setFav="setFav"
        @hidePaginator="togglePaginator"
      />
      <div
        v-if="showPaginator"
        class="pages"
      >
        <ButtonComponent
          label="Previous page"
          :disabled="page === 1"
          @click.native="changePage(-1)"
        />
        <ButtonComponent
          label="Next Page"
          :disabled="beerList.length < 25"
          @click.native="changePage(+1)"
        />
      </div>
    </div>
    <transition name="modal">
      <ModalComponent
        v-if="showModal"
        :beer="beerSelected"
        @closeModal="triggerModal"
      />
    </transition>
  </div>
</template>

<script>
import BeerService from './services/beerServices'

import debounce from 'lodash.debounce'
import { lockScroll, unlockScroll } from './helpers'

import Header from './components/HeaderComponent/HeaderComponent'
import ListComponent from './components/BeerListComponent/BeerListComponent'
import ModalComponent from './components/ModalComponent/ModalComponent'
import ButtonComponent from './components/ButtonComponent/ButtonComponent'

export default {
  name: 'BeerCatalogue',

  components: {
    Header,
    ListComponent,
    ModalComponent,
    ButtonComponent
  },

  data () {
    return {
      beerList: [],
      search: '',
      errorSearch: false,
      page: 1,
      params: {},
      showModal: false,
      beerSelected: {},
      favBeers: [],
      showFav: false
    }
  },

  computed: {
    showPaginator () {
      return !this.showFav && (this.page === 1 ? this.beerList.length >= 25 : true)
    },
    beerListShow () {
      const beersToShow = this.showFav ? this.favBeers : this.beerList
      return beersToShow.map(elem => {
        const isFav = this.favBeers.find(fav => fav.id === elem.id)

        return {
          ...elem,
          isFav: Boolean(isFav)
        }
      })
    }
  },

  watch: {
    search: function (newVal, oldVal) {
      this.debouncedSearchTerms(newVal, oldVal)
    }
  },

  created () {
    this.debouncedSearchTerms = debounce(this.searchTerms, 500)
  },

  async mounted () {
    const { data } = await BeerService.getBeerList()

    this.beerList = data
  },

  methods: {
    async searchTerms (newVal, oldVal) {
      const newValue = newVal.trim()
      this.errorSearch = false
      this.page = 1
      if (newValue === '') {
        const { data } = await BeerService.getBeerList({ page: this.page })
        this.beerList = data
        this.params = {}
      } else {
        const newValueArr = newValue.split(' ')
        const params = {}

        newValueArr.forEach(elem => {
          const regName = new RegExp(/^[A-Za-z\s]+$/g)
          const regDate = new RegExp(/[0-9]{2}-[0-9]{4}/g)

          if (regName.test(elem)) params.beer_name = params.beer_name ? `${params.beer_name} ${elem}` : elem
          else if (regDate.test(elem)) params.brewed_after = elem
        })

        if (Object.keys(params).length) {
          this.params = params
          const { data } = await BeerService.getBeerList({
            ...params,
            page: this.page
          })
          this.beerList = data
        } else {
          this.errorSearch = true
        }
      }
    },
    async changePage (index) {
      if (this.page + index < 1) return
      const params = {
        ...this.params,
        page: this.page + index
      }

      this.page = this.page + index

      const { data } = await BeerService.getBeerList(params)

      this.beerList = data
    },

    triggerModal (data) {
      this.showModal = !this.showModal
      if (this.showModal) {
        this.beerSelected = data
        lockScroll()
      } else {
        this.beerSelected = {}
        unlockScroll()
      }
    },

    setFav (data) {
      const indexFav = this.favBeers.findIndex(fav => fav.id === data.id)
      if (indexFav < 0) this.favBeers.push(data)
      else this.favBeers.splice(indexFav, 1)
    },

    togglePaginator () {
      this.showFav = !this.showFav
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style lang="scss" src="./styles/index.scss" />
