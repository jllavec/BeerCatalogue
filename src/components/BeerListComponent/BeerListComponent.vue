<template>
  <div
    class="beer-list"
  >
    <div class="beer-list__display-selector">
      <ButtonComponent
        class="button beer-list__gallery-trigger"
        :label="gallery ? 'Show List' : 'Show Gallery'"
        @click.native="changeDisplayMode"
      />
      <ButtonComponent
        :label="favShow ? 'Hide Favs' : 'Show Favs'"
        class="button"
        @click.native="showFavs"
      />
    </div>
    <ul :class="{ 'beer-list--gallery': gallery }">
      <li
        v-for="(beer, index) in beerList"
        :key="`beer-${index}`"
        class="beer-list__element"
        @click="showDetail(beer)"
      >
        <p class="beer-list__element-title">
          Name: {{ beer.name }}
        </p>
        <p>
          Tagline: {{ beer.tagline }}
        </p>
        <p>
          {{ beer.first_brewed }}
        </p>
        <p class="beer-list__element-emoji">
          {{ alcoholEmoji(beer) }}
        </p>
        <Star
          class="beer-list__element-fav"
          :class="{'beer-list__element-fav--selected': beer.isFav}"
          @click.stop="setFav(beer)"
        />
      </li>
    </ul>
  </div>
</template>

<script>
import { alcoholEmoji } from '../../helpers'

import Star from '../../assets/images/Five-pointed_star.svg'

import ButtonComponent from '../ButtonComponent/ButtonComponent'

export default {
  name: 'BeerList',

  components: {
    ButtonComponent,
    Star
  },

  props: {
    beerList: {
      type: Array,
      default: () => []
    }
  },

  data () {
    return {
      search: '',
      gallery: false,
      favShow: false
    }
  },

  methods: {
    alcoholEmoji,
    changeDisplayMode () {
      this.gallery = !this.gallery
    },
    showDetail (element) {
      this.$emit('displayModal', element)
    },
    setFav (beer) {
      this.$emit('setFav', beer)
    },
    showFavs () {
      this.favShow = !this.favShow
      this.$emit('hidePaginator', this.favShow)
    }
  }
}
</script>

<style lang="scss" scoped src="./BeerListComponent.scss"  />
