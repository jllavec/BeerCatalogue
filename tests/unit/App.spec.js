import { shallowMount } from '@vue/test-utils'
import App from '@/App.vue'
import http from '@/http'
import beers from '../mocking/beers.json'
import beersPopulated from '../mocking/beersPopulated.json'
import BeerListComponent from '@/components/BeerListComponent/BeerListComponent'
import ButtonComponent from '@/components/ButtonComponent/ButtonComponent'

jest.mock('@/http/index.js')

describe('App.vue', () => {
  describe('-- With empty vals --', () => {
    let wrapper

    beforeEach(() => {
      http.get.mockImplementation(() => Promise.resolve({ data: [] }))
      wrapper = shallowMount(App)
    })

    it('renders component correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('shouldnt show paginator since showPaginator is false', () => {
      expect(wrapper.vm.showPaginator).toBeFalsy()
    })
  })

  describe('-- With populated vals --', () => {
    let wrapper

    beforeEach(() => {
      http.mockReset()
      http.get.mockReset()
      http.get.mockImplementation(() => Promise.resolve({ data: beers }))
      wrapper = shallowMount(App)
    })

    it('renders component correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should show paginator since showPaginator is true', () => {
      expect(wrapper.vm.showPaginator).toBeTruthy()
    })

    it('shouldnt show paginator since showFav is true & showPaginator is false', () => {
      wrapper.setData({ showFav: true })
      expect(wrapper.vm.showPaginator).toBeFalsy()
    })

    it('should show paginator since page is > 1 & showPaginator is true', () => {
      wrapper.setData({ page: 2 })
      expect(wrapper.vm.showPaginator).toBeTruthy()
    })

    it('should map correctly fav beers and set isFav prop properly', async () => {
      const favBeers = beers.slice(0, 5)
      wrapper.setData({ favBeers })
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.beerListShow).toEqual(beersPopulated)
    })

    it('should return just fav list beers when showFav is true', async () => {
      const favBeers = beers.slice(0, 5)
      wrapper.setData({ favBeers, showFav: true })
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.beerListShow).toEqual(favBeers.map(elem => ({ ...elem, isFav: true })))
    })

    it('should call correctly to the api and retrieve data when searchTerm name is properly defined', () => {
      wrapper.vm.searchTerms('lager')
      expect(http.get).toHaveBeenNthCalledWith(2, '/beers', { params: { beer_name: 'lager', page: 1 } })
      expect(wrapper.vm.params).toEqual({ beer_name: 'lager' })
    })

    it('should call correctly to the api and retrieve data when searchTerm name is properly defined', () => {
      wrapper.vm.searchTerms('lager 10-2010')
      expect(http.get).toHaveBeenNthCalledWith(2, '/beers', { params: { beer_name: 'lager', brewed_after: '10-2010', page: 1 } })
      expect(wrapper.vm.params).toEqual({ beer_name: 'lager', brewed_after: '10-2010' })
    })

    it('should throw error when term isnt valid', () => {
      http.get.mockReset()
      wrapper.vm.searchTerms('1s0-2fdf010')
      expect(wrapper.vm.errorSearch).toBeTruthy()
    })

    it('should call to raw page one when search terms are empty', () => {
      wrapper.vm.searchTerms('')
      expect(http.get).toHaveBeenNthCalledWith(2, '/beers', { params: { page: 1 } })
      expect(wrapper.vm.params).toEqual({})
    })

    it('should always reset to page 1 when terms are changed', () => {
      wrapper.setData({ page: 3 })
      wrapper.vm.searchTerms('sdf')
      expect(http.get).toHaveBeenNthCalledWith(2, '/beers', { params: { page: 1, beer_name: 'sdf' } })
      expect(wrapper.vm.page).toBe(1)
    })

    it('should trigger watcher when search data changes', async () => {
      const spyDebounce = jest.spyOn(wrapper.vm, 'debouncedSearchTerms')
      wrapper.setData({ search: 'lager' })
      await wrapper.vm.$nextTick()
      expect(spyDebounce).toHaveBeenCalled()
    })

    it('should change page and call to api when changePage method called', () => {
      wrapper.vm.changePage(1)
      expect(wrapper.vm.page).toBe(2)
      expect(http.get).toHaveBeenNthCalledWith(2, '/beers', { params: { page: 2 } })
    })

    it('shouldnt do anything when trying to navigate to page less than 1', () => {
      http.get.mockReset()
      wrapper.vm.changePage(-1)
      expect(wrapper.vm.page).toBe(1)
      expect(http.get).toHaveBeenCalledTimes(0)
    })

    it('list emitting events should trigger event functions', async () => {
      wrapper.setMethods({
        togglePaginator: jest.fn(),
        setFav: jest.fn(),
        triggerModal: jest.fn()
      })
      const list = wrapper.findComponent(BeerListComponent)

      await list.vm.$emit('hidePaginator')
      await list.vm.$emit('setFav')
      await list.vm.$emit('displayModal')

      expect(wrapper.vm.togglePaginator).toHaveBeenCalled()
      expect(wrapper.vm.setFav).toHaveBeenCalled()
      expect(wrapper.vm.triggerModal).toHaveBeenCalled()
    })
  })
})
