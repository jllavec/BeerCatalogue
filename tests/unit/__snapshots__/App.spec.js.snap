// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue -- With empty vals -- renders component correctly 1`] = `
<div class="app">
  <header-stub></header-stub>
  <div class="container">
    <div class="search"><label for="Name">Search your beer</label> <input id="Name" type="text" name="Name" placeholder="Search by name / brew date (mm-yyyy)">
      <!---->
    </div>
    <listcomponent-stub beerlist=""></listcomponent-stub>
    <!---->
  </div>
  <transition-stub name="modal">
    <!---->
  </transition-stub>
</div>
`;

exports[`App.vue -- With populated vals -- renders component correctly 1`] = `
<div class="app">
  <header-stub></header-stub>
  <div class="container">
    <div class="search"><label for="Name">Search your beer</label> <input id="Name" type="text" name="Name" placeholder="Search by name / brew date (mm-yyyy)">
      <!---->
    </div>
    <listcomponent-stub beerlist="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"></listcomponent-stub>
    <div class="pages">
      <buttoncomponent-stub label="Previous page" disabled="true"></buttoncomponent-stub>
      <buttoncomponent-stub label="Next Page"></buttoncomponent-stub>
    </div>
  </div>
  <transition-stub name="modal">
    <!---->
  </transition-stub>
</div>
`;
